### DISCUSSION TITLE
ESP32 reading data from SPS30 PM Sensor

### DISCUSSION THREAD

**Discussion by And1DS** [micropython/micropython#13263] [Hardware & Peripherals]
Hi, 
I am trying to read some data from the sensirion SPS30 Sensor via i2c. I can find the sensor on the i2c bus with address 0x69. 
The sensor documentation says that there are different transfer types:

```
Set Pointer
Sets the 16-bit address pointer without writing data to the sensor module. It is used to execute commands, which do not
require additional parameters.

Set Pointer & Read Data
Sets the 16-bit address pointer and read data from sensor module. It is used to read sensor module information or
measurement results. The data is ready to read immediately after the address pointer is set. The sensor module transmits
the data in 2-byte packets, which are protected with a checksum.
```

To start with i would like to just read some device information. The serial can be retrieved with pointer address 0xD033

```
Read Device Information (0xD025, 0xD033)
This command returns the requested device information. It is defined as a string value with a maximum length of 32 ASCII
characters (including terminating null-character)
```
According to the doc the read data will return 48 bytes, with Byte# 0,1 containing an ASCII Character, followed by a checksum, Byte# 45,46 containing ASCII caharacter 30 and 31, followed by checksum for bytes 45,46
In my program i use the following code: 

```
import machine

i2c = machine.SoftI2C(sda=machine.Pin(13), scl=machine.Pin(16))
device_address = 0x69
data = bytearray(48)
i2c.writeto(device_address, b'\xD0\x33')
i2c.readfrom_into(device_address, data)
```

when i print data i get the following output: 

`bytearray(b'\xff\xe1C\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff')`

I am not sure if this is the correct way i am supposed to write/read from the i2c? And unfortunately i wasn't able to translate the data back to a readable string which would make sense, sorry for the noob question, can someone point me in to the right direction how to read data from this sensor? thank you

On page 11 of this PDF there is the detailled description of the sensor protocol: https://cdn.sparkfun.com/assets/2/d/2/a/6/Sensirion_SPS30_Particulate_Matter_Sensor_v0.9_D1__1_.pdf

Thanks very much in advance, any help is appreciated!

**Comment by robert-hh** [comment]
If you did not do so, you can make a simple check for the I2C connection:
- check that SDA, SCL, GND and Vcc are connected
- check that there are pull-up resistors on SDA and SCL
- run the commands
```
import machine
i2c = machine.SoftI2C(sda=machine.Pin(13), scl=machine.Pin(16))
print(i2c.scan())
```
It should return a short list containing the number 105 == 0x69.

**Reply by And1DS** [reply]
Hi Robert, 
thank you for your answer. Indeed i2c.scan finds the device, the communication seems to work. I am just not sure how to deal with the data / whats the correct way to interface with the sensor...

**Reply by bidrohini** [reply]
The checksum is an important part of the communication. You need to calculate it correctly to ensure data integrity. The checksum calculation is mentioned in the document you provided.
You can try this code instead: 
`import machine

i2c = machine.SoftI2C(sda=machine.Pin(13), scl=machine.Pin(16))
device_address = 0x69

# Set Pointer to 0xD033
i2c.writeto(device_address, b'\x00\x10\xD0\x33')

# Read Data
data = bytearray(48)
i2c.readfrom_into(device_address, data)

# Checksum Calculation (adjust based on your specific implementation)
checksum = sum(data[:-2]) & 0xFFFF

# Verify Checksum (adjust based on your specific implementation)
if checksum == ((data[-2] << 8) + data[-1]):
    # Convert ASCII characters to a readable string
    result_string = data[2:-2].decode('utf-8')

    print("Device Information:", result_string)
else:
    print("Checksum verification failed. Data may be corrupted.")
`
I hope your sensor will work. You can also try SDS011 and compare the performances. 

https://www.pcbway.com/project/shareproject/ESP8266_Air_Quality_monitor_with_Nova_PM_Sensor_SDS011_888bf8c3.html

**Reply by And1DS** [reply]
Thank you bidrohini for your help and example code, 

I eventually managed to get the sensor working, turned out it was the way i passed the address in i2c.writeto, if anyone has the same issue - the correct way to do it for this sensor in micropython is: 

```
SERIAL_CMD_ADDR =  [0xD0, 0x33]
i2c.writeto(device_address, bytearray(SERIAL_CMD_ADDR)
```

reading PM data from SPS30 now using micropython on an esp32! yay!
